runtime/bcutil/ROMClassBuilder.cpp

```diff
@@ -270,13 +270,32 @@ ROMClassBuilder::handleAnonClassName(J9CfrClassFile *classfile, bool *isLambda,
 	U_32 i = 0;
 	BOOLEAN stringOrNASReferenceToClassName = FALSE;
 	BOOLEAN newCPEntry = TRUE;
+	bool isLambdaForm = false;
 	BuildResult result = OK;
 	char buf[ROM_ADDRESS_LENGTH + 1] = {0};
 	U_8 *hostPackageName = context->hostPackageName();
 	UDATA hostPackageLength = context->hostPackageLength();
 	PORT_ACCESS_FROM_PORT(_portLibrary);
 
-#if defined(J9VM_OPT_OPENJDK_METHODHANDLE) && (JAVA_SPEC_VERSION >= 15)
+#if defined(J9VM_OPT_OPENJDK_METHODHANDLE)
+	/*
+	 * Prevent generated LambdaForm classes from MethodHandles to be stored to the shared cache.
+	 * When there are a large number of such classes in the shared cache, they trigger a lot of class comparisons.
+	 * Performance can be much worse (compared to shared cache turned off).
+	 */
+	if (context->isClassAnon() && context->isClassUnsafe()) {
```

Comment:
The original check was `J9_ARE_ANY_BITS_SET(*options, J9_FINDCLASS_FLAG_HIDDEN | J9_FINDCLASS_FLAG_UNSAFE)`. It is now isClassAnon && isClassUnsafe. Are all the lambdaForm classes both anon and unsafe ?

Reply:
From my observations, all the `LambdaForm`s were Anon and Unsafe.

Reply:
OK. Probably you can add a comment in the code that `LambdaForms are both Anon and Unsafe`.

Reply:
- in jdk8 and jdk11, they should be unsafe/anonoymous. all classes created using `unsafe.defineclass` should be `anonymous`. so, we don't need `anon && unsafe`; just one should suffice. 
- in jdk17+, they are hidden. so, the current change will reintroduce the bug fixed by https://github.com/eclipse-openj9/openj9/pull/13644 in JDK17+.

Reply:
For 17, the `LambdaForms` were Anon, Unsafe, and Hidden at the point of the check.

Reply:
iirc, at some point after JDK17, the usage of `Unsafe.defineAnonymousClass` was stopped for creating `Hidden` classes. `Anon` is only added for weak hidden classes. For strong hidden classes, `Anon` is not added. The current check might miss strong hidden classes which don't have the `Anon` flag. For fully coverage of hidden classes, we should retain the original check: `isClassHidden() || isClassUnsafe()`.

Reply:
`Unsafe.defineAnonymousClass` is not in the backtrace at all for JDK17. All the `LambdaForms` I've observed are like this `java/lang/invoke/LambdaForm$DMH isHidden: 1, isUnsafe: 1, isAnonymous: 1` at the point of the check.

If we allow `isClassHidden() || isClassUnsafe()` we will also `memcmp` against all of the `Lambda` classes.

I'll look to see if there is a more precise set of flags for 17+ we can use so we can avoid checking against `Lambdas` and don't miss the strong hidden classes.

Reply:
What about `isClassAnon() || isHiddenClassOptStrongSet()`?

Reply:
We'd better be more general (less strict) on this `if` check here, as we do not control the JCL code,  which might be changed in the future. The following class name comparison will always find us the lamdaForm classes.

Reply:
> What about isClassAnon() || isHiddenClassOptStrongSet()?

It seems equivalent to `isClassHidden() || isClassUnsafe()`. https://github.com/eclipse-openj9/openj9/blob/0318ff92766bbbdf36c10a202b08b45ef1b6522f/runtime/jcl/common/clsldr.cpp#L108-L123

> we can use so we can avoid checking against Lambdas and don't miss the strong hidden classes.

The above might not be possible.

Reply:
I will just remove that outer if statement entirely.

Reply:
It's correct since `handleAnonClassName` is only invoked for `context->isClassAnon() || context->isClassHidden()` which should cover all `LambdaForm` cases. @hangshao0 Do you agree?

https://github.com/eclipse-openj9/openj9/blob/8c73b1d1aa950a2afc1bf1787488e0b390c0c432/runtime/bcutil/ROMClassBuilder.cpp#L525-L530

Reply:
Exactly.

Reply:
> It's correct since handleAnonClassName is only invoked for context->isClassAnon() || context->isClassHidden() which should cover all LambdaForm cases. @hangshao0 Do you agree?

Yes.